typing/subst.ml

```diff
@@ -54,8 +56,13 @@ let add_modtype id ty s = add_modtype_path (Pident id) ty s
 
 let for_saving s = { s with for_saving = true }
 
+let change_locs s loc = { s with loc = Some loc }
+
 let loc s x =
-  if s.for_saving && not !Clflags.keep_locs then Location.none else x
+  match s.loc with
+  | Some l -> l
+  | None ->
+    if s.for_saving && not !Clflags.keep_locs then Location.none else x
```

Comment:
Is this the right ordering of tests? If I understand correctly, the idea of the `keep-locs` check is to keep `cmi` location-free, so why do we not return `Location.none` in this case even when the substitution has a location?

Reply:
These features do not interact. `change_locs` is only used when producing new signatures, while `for_saving` is (theoretically) used only when saving a signature to a `cmi`.

Reply:
If the two orderings are equivalent, I would maybe have a preference for the one that lets me easily reason about the fact that I will get `Location.none` as expected -- the other order. But okay for having things as they are now.